### ISSUE TITLE
Warn about trait bounds on struct and enum type parameters

### ISSUE THREAD

**Issue by dtolnay** [Manishearth/rust-clippy#1689]
Traits are for behavior.

```rust
impl<T> Data<T> where T: Behavior { /* behavior, okay */ }

fn behave<T>() where T: Behavior { /* also behavior, also okay */ }

struct Data<T> where T: Behavior { /* DON'T DO THIS */ }
```

The only exception is things like [`Cow`](https://doc.rust-lang.org/std/borrow/enum.Cow.html) that use associated types to define data.

My current understanding is that any data structure not using associated types should not have trait bounds (?Sized doesn't count). Redundant bounds like this suck because they transitively infect anything that names the type in any way. Consider [this enum](https://github.com/lawliet89/biscuit/blob/0fa381dcca60cc441b68f59705b43b4baeb158bf/src/jwe.rs#L275-L290).

```rust
#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
enum Compact<T: CompactPart, H: Serialize + Deserialize + Clone + 'static> {
    Decrypted {
        header: Header<H>,
        payload: T,
    },
    Encrypted(Encrypted),
}

// kill me
fn print_the_thing<T, H>(c: Compact<T, H>)
    where T: CompactPart + Debug,
          H: Serialize + Deserialize + Clone + 'static + Debug
{
    println!("{:?}", c);
}
```

Without trait bounds on the data structure:

```rust
fn print_the_thing<T: Debug, H: Debug>(c: Compact<T, H>) {
    println!("{:?}", c);
}
```

**Comment by dtolnay** [comment]
@oli-obk I think this would help a lot with people trying to write Deserialize\<'de> bounds on their types. Almost all the struggles I've seen have been on structs, not so much on functions, and the answer is always just don't put a trait bound there.

**Comment by Manishearth** [comment]
Hm, this is an interesting lint. Unsure if it should be warn by default; probably should.

**Comment by dtolnay** [comment]
I believe it should warn by default as long as we can reliably detect the use of an associated type.

**Comment by clarcharr** [comment]
Fun fact: a few libstd types could benefit from these changes.

**Comment by dtolnay** [comment]
Which ones did you find? I found:

- [`std::slice::Split`](https://doc.rust-lang.org/std/slice/struct.Split.html)
- [`std::slice::SplitMut`](https://doc.rust-lang.org/std/slice/struct.SplitMut.html)
- [`std::slice::SplitN`](https://doc.rust-lang.org/std/slice/struct.SplitN.html)
- [`std::slice::SplitNMut`](https://doc.rust-lang.org/std/slice/struct.SplitNMut.html)
- [`std::slice::RSplitN`](https://doc.rust-lang.org/std/slice/struct.RSplitN.html)
- [`std::slice::RSplitNMut`](https://doc.rust-lang.org/std/slice/struct.RSplitNMut.html)
- [`std::collections::binary_heap::BinaryHeapPlace`](https://doc.rust-lang.org/std/collections/binary_heap/struct.BinaryHeapPlace.html)

**Comment by dtolnay** [comment]
For the longest time before 2013 https://github.com/rust-lang/rust/commit/6ce74460e6a5c8045a7b43b86a656f28354f4b0c#diff-76432ae7ccf0520cf36aac7c006809a3 :wink:

**Comment by oli-obk** [comment]
Won't the error messages quality decrease if those bounds are as late as possible? Especially for iterator adaptors? Don't know about any of the conversations on this topic, so I might be mixing some. A specialized version for a whitelist of bounds would probably be a good start. Things like debug, partialeq, deserialize...

**Comment by llogiq** [comment]
I'd also like to see how such a lint would fare against [typenum](https://github.com/paholg/typenum).

**Comment by dtolnay** [comment]
> I'd also like to see how such a lint would fare against [typenum](https://github.com/paholg/typenum).

Perfectly fine. I tried removing all trait bounds from structs in typenum and it still compiles and passes all tests.

**Comment by dtolnay** [comment]
> Won't the error messages quality decrease if those bounds are as late as possible? Especially for iterator adaptors?

@oli-obk could you put together an example with before and after code? I don't think this could happen. Errors are triggered when you rely on behavior, and we would be keeping trait bounds on behavior.

**Comment by llogiq** [comment]
cc @paholg

**Comment by paholg** [comment]
It's beyond the scope of clippy, but I would like to see the opposite; that putting a constraint on a struct implicitly adds that constraint to all impls of that struct. This would mean that removing constraints would become a potentially breaking change, though.

As is, removing constraints from structs shouldn't break things, but it moves information on how the struct should be used from the type signature into documentation.

**Comment by dtolnay** [comment]
Constraints on a struct implying the same constraints on all impls is an orthogonal discussion. Even with that feature, I would strongly disagree with putting trait bounds on data and I think this lint would still be valuable.

Consider Vec\<T\>. You can clone a vec if you can clone T. You can serialize a vec if you can serialize T. You can deserialize a vec if you can deserialize T.

Look at the code I linked to in the top comment. I see this all the time. It is the equivalent of writing:

```rust
#[derive(Clone, Serialize, Deserialize)]
struct Vec<T> where T: Clone + Serialize + Deserialize { /* ... */ }
```

Yes your crate may compile either way. Yes "information on how the struct should be used [is in] the type signature [not] documentation." But no it's not valuable in any way.

**Comment by paholg** [comment]
It's not valuable for a general purpose struct like `Vec`, but for something like typenum, it makes sense. The structs defined in typenum are not designed to have any parameters except those defined inside the crate.

I think this lint could be valuable, too, and I should have expressed that in the previous comment. It's easy enough to disable in the rare cases when you don't want it.

**Comment by dtolnay** [comment]
Makes sense. On a spectrum between Vec\<T\> and typenum::PInt\<U\>, where do you think the Compact\<T, H\> type stands? And as a fraction of types how common do you think the typenum situation of opting out of this lint would be?

**Comment by llogiq** [comment]
I think the typenum case isn't so common – indeed I hope it goes away sooner than later, once Rust's type system can express integers directly.

Another crate I know that uses generic constraints a lot is diesel. Perhaps @killercup can tell us about their uses?

**Comment by killercup** [comment]
I haven't read the whole issue thread, but let me chime in regardless :)

I'm pretty sure that since most _public_ data structures contain some private fields (and they typically should for forwards compat reasons), they also have at least one associated constructor method (usually `new`). Adding constraints to this impl might not be as descriptive as adding them to the struct, but still results in the same thing.

As far as I can ~~see~~ grep, diesel's main crate only has a handful of type constraints used in structs, and only [`RawBytesBindCollector`](http://docs.diesel.rs/diesel/query_builder/bind_collector/struct.RawBytesBindCollector.html) and [`Nullable`](http://docs.diesel.rs/diesel/types/struct.Nullable.html) are public. (And Nullable uses it to restrict recursion.)

As for private structs, it's possible/easier to just use the type constructors, so you need to keep track of the intended constraints yourself, or add them to your struct.

**Comment by kennytm** [comment]
This lint would be pointless if (big if) [implied bounds](https://internals.rust-lang.org/t/lang-team-minutes-implied-bounds/4905) is solved.

**Comment by dtolnay** [comment]
As discussed in https://github.com/Manishearth/rust-clippy/issues/1689#issuecomment-296396580, this lint would be valuable whether or not implied bounds are solved. The typenum-like cases where you would want to opt out are rare.

**Comment by kennytm** [comment]
@dtolnay It think it makes sense for structures like `HashMap<K, V>`, which is useless unless `K: Eq + Hash`. By the time implied-bounds is really landed the lint could be changed to allow-by-default.

**Comment by dtolnay** [comment]
I would disagree with allow-by-default even after implied bounds have landed.

Let's look at the motivating example again. Keep in mind that this is real code written by a real live Rust user of the sort that would rely on Clippy for guidance in navigating the complicated world of Rust types.

```rust
#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
enum Compact<T: CompactPart, H: Serialize + Deserialize + Clone + 'static> {
    Decrypted {
        header: Header<H>,
        payload: T,
    },
    Encrypted(Encrypted),
}
```

Oh but we'd like to add a PartialOrd impl because it is needed for some new code. Easy enough:

```diff
- #[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
- enum Compact<T: CompactPart, H: Serialize + Deserialize + Clone + 'static> {
+ #[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize, PartialOrd)]
+ enum Compact<T: CompactPart, H: Serialize + Deserialize + Clone + PartialOrd + 'static> {
```

Boom, breaking change. Whether or not there are implied bounds. In general this pattern conflicts with the ways we like to think about adding functionality to data structures.

The HashMap case is what we mean by the rare exceptions. Everyday code is much more like the code in this comment.

**Comment by kennytm** [comment]
I disagree that HashMap is a rare exception. 

Even in your example, while `H: Serialize + Deserialize + Clone + 'static` is bad, I don't think the type is useful without `T: CompactPart`.

Let's check more examples (just chosen from the front page of crates.io):

* `quick_xml`: [`struct Reader<B: BufRead>`](https://docs.rs/quick-xml/0.7.1/quick_xml/reader/struct.Reader.html) (Is the type useful only if `B` is `BufRead`? — yes)
* `calamine`: [`struct Cell<T: Default + Clone + PartialEq>`](https://docs.rs/calamine/0.11.2/calamine/struct.Cell.html) (Is the type useful only if `T` is `Default+Clone+PartialEq`? — no)
* `timely`: [`struct Stream<S: Scope, D>`](http://www.frankmcsherry.org/timely-dataflow/timely/dataflow/stream/struct.Stream.html) (`Scope` has associated type, so doesn't count. `D: Data` given in documentation. Is the type useful only if `D: Data`? — likely)
* `textreader`: [`struct TextReader<R: Read>`](https://gyu-don.github.io/rust-textstream/textstream/struct.TextReader.html) (Is the type useful only if `R` is `Read`? — yes)
* `smallstring`: [`struct SmallString<B: Array<Item=u8> = [u8; 8]>`](https://docs.rs/smallstring/0.1.1/smallstring/struct.SmallString.html) (`Array` has associated type, so doesn't count.)


One problem out of four would mean the lint is valuable, but that doesn't mean turning off the lint is rare exception either. What I can say is, the "no lint" condition needs to be much broader than "using associated type" if this is needs to be kept warn-by-default after implied-bounds.

**Comment by dtolnay** [comment]
I would settle for deny-by-default on a blacklist of traits. Namely:

- Clone
- PartialEq
- PartialOrd
- Debug
- Display
- Default
- Serialize
- Deserialize
- DeserializeOwned

I think this would catch a high enough fraction of actual misuses that the people misusing these trait bounds would be able to extrapolate to other trait bounds they are using in similar ways.

**Comment by kennytm** [comment]
Sounds good to me :)

**Comment by devyn** [comment]
Following this has generally made my code nicer, but I have found one little hiccup that I'm not quite sure what to do about.

Including trait bounds in the data definition allows access to associated items inside traits. This is often done with traits such as `Iterator`, `Future`, and `Stream`.

In my case I'm writing a stream adaptor that contains a buffer of items, but I think Peekable is a good enough analogy:

```rust
pub struct Peekable<S: Stream> {
    stream: Fuse<S>,
    peeked: Option<S::Item>,
}
```

This is taken from [peek.rs](https://docs.rs/futures/0.1.28/src/futures/stream/peek.rs.html#11-14) in futures.

If we want to write this without the `S: Stream` bound, our only option is:

```rust
pub struct Peekable<S, T> {
    stream: Fuse<S>,
    peeked: Option<T>,
}
```

and then to include a `where S: Stream<Item=T>` bound on the implementations. However this makes the types when fully written out a lot longer and potentially include duplication of `T` at many different layers. If you're in a situation where automatic typing is not an issue, this is probably okay although leads to less than ideal errors. But if ever anyone has to write out multiple layers of this, it's a real problem.

However, including the `S: Stream` bound on the struct is also problematic for the same sort of reason others have mentioned: if there are excessive trait bounds on the implementation of `Stream` for a given type you want to use with it inside another structure, you have to make sure that structure *also* has all of those required as trait bounds in some way. So you may run into the same problem eventually.
